machine/v1/types_alibabaprovider.go

```diff
@@ -0,0 +1,340 @@
+	// PL3: A single ESSD can deliver up to 1,000,000 random read/write IOPS.
+	// For more information about ESSD performance levels, see ESSDs.
+	// +kubebuilder:validation:Enum="PL0"; "PL1"; "PL2"; "PL3"
+	// +optional
+	PerformanceLevel AlibabaDiskPerformanceLevel `name:"performanceLevel,omitempty"`
+
+	// Category describes the type of data disk N.
+	// Valid values:
+	// cloud_efficiency: ultra disk
+	// cloud_ssd: standard SSD
+	// cloud_essd: ESSD
+	// cloud: basic disk
+	// Empty value means no opinion and the platform chooses the a default, which is subject to change over time.
+	// Currently for non-I/O optimized instances of retired instance types, the default is `cloud`.
+	// Currently for other instances, the default is `cloud_efficiency`.
+	// +kubebuilder:validation:Enum="cloud_efficiency"; "cloud_ssd"; "cloud_essd"; "cloud"
+	// +optional
+	Category AlibabaDiskCategory `name:"category,omitempty"`
+
+	// KMSKeyID is the ID of the Key Management Service (KMS) key to be used by data disk N.
+	// Empty value means no opinion and the platform chooses the a default, which is subject to change over time.
+	// Currently the default is `""` which is interpreted as do not use KMSKey encryption.
+	// +optional
+	KMSKeyID string `name:"kmsKeyId,omitempty"`
+
+	// DiskPreservation specifies whether to release data disk N along with the instance.
+	// Empty value means no opinion and the platform chooses the a default, which is subject to change over time.
+	// Currently the default is `DeleteWithInstance`
+	// +kubebuilder:validation:Enum="DeleteWithInstance";"PreserveDisk"
+	// +optional
+	DiskPreservation AlibabaDiskPreservationPolicy `name:"diskPreservation,omitempty"`
+}
+
+// Tag  The tags of ECS Instance
+type Tag struct {
+	// Key is the name of the key pair
+	Key string `name:"Key"`
+	// Value is the value or data of the key pair
+	Value string `name:"value"`
+}
+
+// Bandwidth describes the bandwidth strategy for the network of the instance
+type BandwidthProperties struct {
+	// InternetMaxBandwidthIn is the maximum inbound public bandwidth. Unit: Mbit/s. Valid values:
+	// When the purchased outbound public bandwidth is less than or equal to 10 Mbit/s, the valid values of this parameter are 1 to 10.
+	// Currently the default is `10` when outbound bandwidth is less than or equal to 10 Mbit/s.
+	// When the purchased outbound public bandwidth is greater than 10, the valid values are 1 to the InternetMaxBandwidthOut value.
+	// Currently the default is the value used for `InternetMaxBandwidthOut` when outbound public bandwidth is greater than 10.
+	// +optional
+	InternetMaxBandwidthIn int64 `json:"internetMaxBandwidthIn,omitempty"`
+
+	// InternetMaxBandwidthOut is the maximum outbound public bandwidth. Unit: Mbit/s. Valid values: 0 to 100.
+	// Empty value means no opinion and the platform chooses the a default, which is subject to change over time.
+	// Currently the default is `0`
```

Comment:
by default we disallow egress traffic?

Reply:
The documentation here is lacking. I looked up this information during the latst review but was unable to find something more concrete. According to the docs https://www.alibabacloud.com/help/doc-detail/25499.htm it states
```
	
The maximum outbound public bandwidth. Unit: Mbit/s. Valid values: 0 to 100.

Default value: 0.
```
Another explanation states the following (https://partners-intl.aliyun.com/help/doc-detail/51198.html?spm=a3c0i.10721930.0.0.783e3d98N6R3k4)
```
      Set internet output bandwidth of instance. Unit is Mbps(Mega bit per
      second). Range is [0,200]. Default is 1.While the property is not 0,
      public ip will be assigned for instance.
```
I will ask Alibaba to clarify what the default behavior is on this.

@menglingwei @bd233

Reply:
@kwoodson It may be [this document ](https://www.alibabacloud.com/help/en/doc-detail/63440.html).
The api name is RunInstances

![image](https://user-images.githubusercontent.com/2653769/145313116-b84ad958-5745-4ebb-9847-45e156203f61.png)

![image](https://user-images.githubusercontent.com/2653769/145313148-89ba8f68-0476-4bb8-a6d0-98e8b0f6f039.png)

Reply:
@kwoodson

Reply:
Why would I want to set a value that is not maximal for this? If I want an assigned egress IP, why would I limit the traffic? Is it for cost reasons?